src/LinearMath/btIDebugDraw.h

```diff
@@ -166,9 +166,9 @@ class	btIDebugDraw
 	virtual void drawTransform(const btTransform& transform, btScalar orthoLen)
 	{
 		btVector3 start = transform.getOrigin();
-		drawLine(start, start+transform.getBasis() * btVector3(orthoLen, 0, 0), btVector3(1.f,0.3,0.3));
-		drawLine(start, start+transform.getBasis() * btVector3(0, orthoLen, 0), btVector3(0.3,1.f, 0.3));
-		drawLine(start, start+transform.getBasis() * btVector3(0, 0, orthoLen), btVector3(0.3, 0.3,1.f));
```

Comment:
The old lines above look like a typo, to me.  All of the Bullet code I've seen uses double format for literal numerical values.

That said... if Bullet were to use 32-bit float formatting for all literals then most of the casts to btScalar would be unnecessary, since the compiler would silently upcast to double when necessary.

Reply:
Actually, 32 bit float types can EXACTLY represent integer values up to around 8 million (and down to around -8 million). This is because 32 bit floats have 23 bits of mantissa, so rounding doesn't start to occur for integer values until 2^23 (~8 million).

So in fact,
 float(1.0) == double(1.0)

But,
 float(0.1) != double(0.1)

This is so because 0.1, while it is exactly represented in base-10, it has no exact representation in base-2. Much like how the fraction 1/9 cannot be represented exactly in base-10, but in base-9 it can.

Reply:
may be use "User-defined literals" from c++11